site/content/blog/2020-06-10-opentracing.md

```diff
@@ -0,0 +1,576 @@
+	// execute the template and pass it the metadata we collected before
+	t.Execute(out, metadata)
+	// add needed imports and format the code before printing
+	formattedCode, _ := imports.Process("animal_tracer.go", out.Bytes(), &imports.Options{Comments: true})
+	// print it out!
+	fmt.Println(string(formattedCode))
+```
+
+The result will be:
+
+```go
+package animals
+
+import "fmt"
+
+type AnimalTracer struct {
+	IAnimal
+}
+
+func (a *AnimalTracer) Speak(x int) {
+	fmt.Printf("Running Speak(x) with 'x'=%v ", x)
+	a.IAnimal.Speak(x)
+}
+```
+
+Beautiful!
+
+Similarly, running the generator over the `timerTemplate` will yield:
+
+```go
+package animals
+
+import (
+	"fmt"
+	"time"
+)
+
+type AnimalTimer struct {
+	IAnimal
+}
+
+func (a *AnimalTimer) Speak(x int) {
+	fmt.Println("Timing Speak function...")
+	__t := time.Now()
+	a.IAnimal.Speak(x)
+	fmt.Printf("Speak took %s\n", x, time.Since(__t))
+}
+```
+
+## Finishing up
+
+Using the techniques from the previous section, we can now generate the Open Tracing decorator we want by using the following template:
+
+```go
+{{range $index, $element := .Methods}}
+func (a *{{$.Name}}) {{$index}}({{$element.Params | joinParamsWithType}}) {{$element.Results | joinResultsForSignature}} {
+	origCtx := a.ctx
+	span, newCtx := tracing.StartSpanWithParentByContext(a.ctx, "app.{{$index}}")
+
+	a.ctx = newCtx
+	a.app.Srv().Store.SetContext(newCtx)
+	defer func() { 
+		a.app.Srv().Store.SetContext(origCtx)
+		a.ctx = origCtx 
+	}()
+	{{range $paramIdx, $param := $element.Params}}
+		{{ shouldTrace $element.ParamsToTrace $param.Name }}
+	{{end}}
+	defer span.Finish()
+	{{- if $element.Results | len | eq 0}}
+		a.app.{{$index}}({{$element.Params | joinParams}})
+	{{else}}
+		{{$element.Results | genResultsVars}} := a.app.{{$index}}({{$element.Params | joinParams}})
+		{{if $element.Results | errorPresent}}
+			if {{$element.Results | errorVar}} != nil {
+				span.LogFields(spanlog.Error({{$element.Results | errorVar}}))
+				ext.Error.Set(span, true)
+			}
+		{{end}}		
+		return {{$element.Results | genResultsVars -}}
+	{{end}}}
+{{end}}
+```
+
+Phew, this was quite a trip, huh? I hope you found it interesting. You can see the actual generator implementation inside `mattermost-server` in [/app/layer_generators/main.go](https://github.com/mattermost/mattermost-server/blob/master/app/layer_generators/main.go)
+
+> __Side note__: this is just one way of handling this problem. Some people are against using code-generation too much since it hides a lot of implementation away and complicates the build process (you have to re-run the generators each time your interface changes). We've settled on this approach due to it's flexibility and performance. If you have any notes or ideas on how this could be implemented cleaner - please stop by the [Mattermost Community](https://community.mattermost.com) and I'll be **very** glad to discuss it further.
```

Comment:
```suggestion
*Side note*: This is just one way of handling this problem. Some people are against using code-generation too much since it hides a lot of implementation and complicates the build process (you have to re-run the generators each time your interface changes). We've settled on this approach due to its flexibility and performance. 

If you have any notes or ideas on how this could be implemented in a cleaner way - please stop by the [Mattermost Community](https://community.mattermost.com) server and I'll be **very** glad to discuss it further.
```